### ISSUE TITLE
How to mimic the login flow for registration...?

### ISSUE THREAD

**Issue by sneko** [wi3land/ionic-appauth#16]
Hi @wi3land ,

Usually when going on the "accounts.XXXX.com" page for the OAuth2 login, companies like Google leave the choice to either login or register or recover the password.

Currently with `ionic-appauth` I'm able to say `open the accounts page for the login and then redirect back the user`. What would be the best approach to set a "register" button inside my app and try to mimic the OAuth2 flow? I mean, I don't want to reinvent the wheel for "Open Registration view" with RedirectURI...

Did you already see that implemented?

An idea I have to hack a bit would be as discussed yesterday in another issue, to use the `signIn()` method from `ionic-appauth` and add a parameter in the state (needs to be implemented), or for now stringify an object and pass it as `loginHint` (just for a test).

In this object I could have something like: `{"action": "register"}`, so the `accounts.XXXX.com` website would know which page to show. And since the challenge code during the registration would be the initial login one, after registration+login the user would be redirected to the app.

What do you think :) ?

**Comment by wi3land** [comment]
you can force additional parameters into the url via the IAuthConfig auth_extras string map. This sends over any additional parameters on authorization call.

Or you could utilise the plugins provided by the library to add the functionality you require like:

```
public register() {
    const returnURL = encodeURI('appauth://registerCallback');
    const url = `http://${this.authConfig.identity_server}/account/register?returnUrl=${returnURL}`;
    this.browser.showWindow(url);
}
```


Then in the handleCallback method you would just need to wait for the register and do whatever you wanted to do on the way back.


```
 private handleCallback(callbackUrl: string): void {
    if ((callbackUrl).indexOf(this.authConfig.redirect_url) === 0) {
      this.AuthorizationCallBack(callbackUrl);
    }

    if ((callbackUrl).indexOf(this.authConfig.end_session_redirect_url) === 0) {
      this.EndSessionCallBack();
    }

    if ((callbackUrl).indexOf('appauth://registerCallback') === 0) {
      //Do Something After Register
      this.authSubject.next({ action: 'Register' });
    }
  }
```

**Comment by sneko** [comment]
Hi @wi3land ,

The problem I see with the second solution is that, if the user is redirected onto the registration flow with  `this.browser.showWindow(url);` (the `register()` method you mentioned), and at the end this one clicks onto "Login" from the registration page (because for X reason he remembers having an account), no login challenge would be set up.

In this case I could try to redirect to my `oauth2/auth` endpoint with the redirectURI, but I would not be aware of the scopes needed from the client. So maybe using `auth_extras` makes more sense even if it implies poluting my database as soon as a user browse registration pages... What do you think?

Overall, I'm still wondering what would be the better flow for a user after submitting the form. Let's say I send him an email to confirm its registration:
- From there, should the link in the email redirects the user to my client application, or onto the `accounts.XXXX.com` pages? (Since the idea is to centralize login/registration/passwordRecovery I guess I should redirect the user onto the `accounts.XXXX.com`)
- Like at the beginning of my comment, since the user comes from an email there is no login code challenge initiated, except if from the registration I pass the login challenge into the email. But does it sounds good?

I'm glad OpenID Connect brings some standards over OAuth2 to manage authentication, but I keep seeing all this stuff really hard to leverage. About my problem I found some articles like https://openid.net/specs/openid-connect-prompt-create-1_0.html but I need to keep reading to find what needs to be done ^^

**Comment by wi3land** [comment]
The annoying thing about the auth_extras is at the moment they are global. So you will have to set the auth_extras on entry to register and make sure you clear them or you will always go to the register flow.

In regards to your question about email confirm redirecting. You are authenticating to the identity server not the mobile app. So email confirmations should be redirected to the identity server. Once that is complete they can login in on the app. My personal opinion is that we shouldn't try to force the openid flow to make user experience easier, this can lead to security flaws.

**Comment by sneko** [comment]
So let's take this example:

1. The user opens the app and does not have an account
2. He clicks on "Sign up" to register
3. AppAuth opens a browser tab (with `auth_extras` with `action=signup`)
4. The user fills the form and submit
5. The user gets the confirmation email and clicks on the link
6. The user gets directed to the `accounts.XXXX.com/confirm?YYY`, that shows him the account is now enabled

But what should happen next?

Either:
- In the confirmation email link I put the login challenge (even if "signup" action), like that the user can can go from `accounts.XXXX.com/confirm?YYY` to `accounts.XXXX.com/login?code=ZZZZ` and if successful he gets redirected into the native application
- Or, in the confirmation email link I put the "client URL" and once the user gets on the confirmation page I ask the user if he wants to come back to the original client application (mobile app, dashboard...), and even if native it works (thanks to "Universal Links"). But once in the client app he will see a new time the browser tab opening to initiate a OpenID Connect login flow. This seems for me the worst experience a user could have, no?
- Another idea?

Thank you,

**Comment by sneko** [comment]
Another question that comes to my mind is:

Examples always show the idea of a end-user coming from a client application. But what if the user goes for any reason to `accounts.XXXX.com` directly? In this case since there is no clientID the login form should never work.

What do you advise to overcome this situation?

I thought I could check if a login challenge code is provided when going onto `accounts.XXXX.com/*` and if none, I start a OpenID Connect login flow with a new created client for the website `accounts.XXXX.com/*`.

After he successfully logs in, the user is still onto `accounts.XXXX.com/* (maybe /profile path)`, can see some information about him, and below I could list all official client applications (mobile app, dashboard) so the user can go where he wants to go.

What do you think?

Sorry for all those questions, if you know a good place for all OIDC discussions (like a private Slack, an old IRC & co) I'm definitely interested in! Grabing some info will help me taking the "wise decision" since it seems no recommended way exists ^^

Thank you,

**Comment by sneko** [comment]
I ended up using `auth_extras` to pass to parameter that it's for registration. I made some modification on my backend so the frontend shows accordingly.

I put the challenge code in all view change (forgot password, login, registration...), but also in the registration confirmation email, like that, as soon as the user logs in the redirect URI will send him back to the original client.

For sure, if the user takes too long, the code will expire, in this case I show him his profile and list of official applications we have with links.

Still curious about other thoughts on this, it was not obvious 😄
